Add Progress Sync Server Support, updates config and logging, Start/Finish dates, and scheduled/automatic sync functionality

jbhul opened the pull request:
Major revisions that greatly expand and revise functionality.

jbhul committed:
Add ProgressSync Support

Allows the KOReader Sync plugin to connect to a ProgressSync Server in order to update relevant metadata within Calibre

Need to test edge cases and clean up config/add options. Fully functional with the intended use case. Should confirm with KOReader ProgressSync Server maintainers to see if this usage is permitted (No EULA). Also NGL the code for logging wasn't super easy for me to read and I'm lazy no idk how well it handles exceptions that I didn't expect.

jbhul committed:
Add Scheduler for ProgressSync

Allows for daily scheduled progress sync by implementing QTimer
Also refactors config slightly to clean up for future ProgressSync settings

Questions
Any way to tell user to restart in order to apply changes using built in calibre functions?
I added a 'addToLayout': True, metadata to the checkboxes to support out of order config settings.
Any good way to show that the service is indeed working? A log of sorts? IDK

Note: I still need to test if it works over multiple days, 1st time run definitely works

jbhul committed:
Add Sync on Device Connect

Adds a setting to sync from KOReader automatically on device connection.

Slight config reorder.

Would be good to implement a setting in sync_to_calibre() that allows it to be silent except for errors to remove the popup.

jbhul committed:
Refactor Dialogs to Table vs det_msg

Implements a custom dialog for the sync logs including a tabular log view for readability.

Note: While functional I think a conversation around what is actually relevant in the user log would be beneficial.

Also implements silent as a bool param for the 3 sync function to not generate a user dialog. Updated automatic device sync and scheduled ProgressSync to be silent

Does not include icon/image for the dialog, can be added in if needed/wanted (info/warn/error/etc)

jbhul committed:
Implement show_restart_warning in config

The settings that require a restart are checked for changes at save and prompt user for restart with a restart button.

Had to split out the automatic device sync from CustomCheckboxLayout to be able to read the value easily. The formatting is off (checkbox left/right, tooltip, etc.) so ideally it goes back into CustomCheckboxLayout, might need some help with that.

jbhul committed:
Adds Semi-Automated Column Creation

**Major refactor of config columns**
Borrows and adapts code from Goodreads Sync to prefill custom columns in KOSync config, making adding the custom columns a breeze.

ReadMe has also been updated with this and past changes.

Goodreads Sync has a GPL v3 license

jbhul committed:
Update Config UI (bugfixes)

Refactors CustomCheckboxLayout similarly to columns from a prior commit.
Refactors ProgressSyncPopup to QDialog
Fixes missing self.must_restart = False

jbhul committed:
Update README.md

Minor update about reading status based on reviewing the code

jbhul committed:
Config bugfixes

Changed is_multiple as (I think) needed. Would benefit from comparing custom column setup data with maintainer(s).
Fixed checkbox saving.

jbhul committed:
Goodreads Sync Extension

Added the ability for this plugin to build on a concurrently installed and configured Goodreads Sync plugin to allow progress changes to be pushed to Goodreads

Should confirm with Goodreads Sync author (and API key holder) to ensure usage is fine.

Does not push reviews to GR yet

jbhul committed:
Adds Start/Finish Dates

Adds started/finished reading columns. Logic based on status, status change dates, and current values for said dates.

Updates ReadMe with said changes

jbhul committed:
Add Goodreads Rating Sync

Allows plugin to automatically push rating to Goodreads upon book completion.

Also:
Fixed deleted line in readme
Minor config formatting cleanups/sectioning/bugfix

jbhul committed:
Revert "Config bugfixes"

This reverts commit 836a3172132d178054b228536f0403cfce445864.

jbhul committed:
Merge pull request #1 from jbhul/Goodreads-Sync

Goodreads sync

jbhul committed:
Progress Sync Only reading

Checks if book with MD5 has status reading and progress < 100 before checking Progress Sync Server (for API efficiency and performance).

Updated comparison operators in some cases

jbhul committed:
ProgressSync, popups, and config update

Added error handling and better logging to progress sync.

Icons are now added to the new SyncCompletionDialog. Copy button has also been added. Slight tweaks to the formatting of the dialog.

If debugging don't be silent in popups for "background" syncs.

Minor config changes.

jbhul committed:
Goodreads Sync Handle No ID

Skips books without goodreads_id when doing Goodreads Sync. Updated readme for goodreads sync

kyxap commented:
Hey! Thanks for that PR! Currently I'm a way from my main dev pc to test it and give the feedback. But I wonder is there a way to split that huge PR (+1,129 −301 ) to multiple smaller ones? It will be much easy to review/test and publish.

jbhul commented:
No worries, the code generally is split into each major change, but bugfixes/tweaks occurred throughout the commits so there isn't really a clear division within them. For example the first commit is a functional feature (ProgressSync Server), but likely has some bugs or changed behavior in a future commit. I think it would be best to kind of evaluate this fork as a major revision to the plugin, there's very little I didn't touch in some way. 

I run windows so I can't test calibre to KOReader (the perms bug from a previous issue that's open). Everything else I tested and is functional, but obviously good to check different environments and/or test further. 

Let me know if you want to hop on discord or something to walk through things or anything I can do to be helpful here.

Also while I did submit the PR I'm working with the Goodreads Dev to make sure the usage I have is chill so please don't merge it until I get the all clear on that front.

I got confirmation from the KOReader folks about the API usage there so all good on that front.

jbhul committed:
ProgressSync Edge Case and Scheduling Fix

Handles edge case where there is no matching progressSync for the MD5. Fixes timer re-scheduling so it actually runs daily.

jbhul committed:
Update .gitignore

jbhul committed:
Allow QLabel to toggle checkbox and disable scroll to change QSpinBox

jbhul committed:
Remove GoodReads Sync Feature

After discussing with the GR calibre plugin dev I decided it wasn't worth it to include the feature for 2 main reasons.
1. Maintainability concerns since this plugin would be dependent on his plugin.
2. This feature falls outside of the scope of this plugin entirely.

jbhul committed:
Convert KOReaderSync to an async thread

Changes sync behavior to not freeze calibre during the sync and instead implement a progress bar and async thread to separate out the calibre UI

jbhul committed:
Update SyncCompletionDialog

Simplifies SyncCompletionDialog and removes custom Icon QWidget

jbhul committed:
Make log cleaner for no changes

Makes no-change items not appear in log unless DEBUG is enabled

jbhul committed:
Refactor CUSTOM_COLUMN_DEFAULTS

CUSTOM_COLUMN_DEFAULTS refactored to change key to be uniform and add features (additional_params, first_in_group, data_source, is_multiple ) and simplify the code.
Updates to dependent code for these features and refactors within config and action.
Custom columns reordered in config for a more organized config UI.

In action.py updated item added to results if user is in DEBUG mode.

jbhul committed:
Organize rows/cols in SyncCompletionDialog

Sorts rows and cols and adds variable header height to SyncCompletionDialog.
Removed book_id from log.

jbhul committed:
Support callbacks during update_metadata

Adds support for callback extensions on sync

jbhul commented:
Ready to merge on my end. 

I removed the GR Sync feature for the reasons described in the commit. Mainly it's far out of scope for this plugin and could present maintainability challenges in the future.

jbhul commented:
@kyxap Howdy, Would it be possible to get a review of this PR?

Let me know if you want to hop on discord or something to walk through things or anything I can do to be helpful here.

kyxap commented on .gitignore:
```diff
@@ -148,3 +148,9 @@ release
+TODO*
+
+# Python Debug Files
+test_*.py
```
Do you have any tests to share? I think this project needs some approach for testing.

Other way I do not think those changes to`.gitignore` file will be beneficial or we can have some sort of dir `./put_your_custom_stuff_here` you can put any temp files there.

jbhul replied:
These two items were effectively a ./put_your_custom_stuff_here for me. test_*.py were so I could test some of the logic w/o being coupled to the plugin (ProgressSync API comes to mind) and I made my own TODO file that I didn't want getting pushed so I added that to the gitignore. These changes can be rolled back before merge if you'd like.

kyxap replied:
will directory `./temp` going to work for you? if so lets add it instead

jbhul replied:
Done

kyxap commented on README.md:
```diff
@@ -155,6 +174,9 @@ them [here](https://github.com/harmtemolder/koreader-calibre-plugin/issues).
   great [Kobo Utilities](https://www.mobileread.com/forums/showthread.php?t=215339)
   calibre plugin.
+- Some code borrowed from--and heavily inspired by--the
+  great [Goodreads Sync](https://www.mobileread.com/forums/showthread.php?t=123281)
```
Are we still have GS here?

jbhul replied:
Code for creating custom cols was borrowed and inspired by the Goodreads Sync plugin even through there is no GS function anymore.

kyxap commented on action.py:
```diff
@@ -165,6 +187,7 @@ def genesis(self):
 
         base = self.interface_action_base_plugin
         self.version = f'{base.name} (v{".".join(map(str, base.version))})'
+        self.EXTENSION_CALLBACK = None
```
If I'm not mistaking in Python, using UPPERCASE names like `EXTENSION_CALLBACK` typically means a constant — something that does not change after it's set

jbhul replied:
Not too sure about the python convention, but this var is defined on line 273 but set to None here just so it's instantiated.

I think it should be lowercase and I can update that.

jbhul replied:
Done

kyxap commented on action.py:
```diff
@@ -488,6 +551,20 @@ def update_metadata(self, uuid, keys_values_to_update):
+        if self.EXTENSION_CALLBACK:
+            try:
+                updateLog = self.EXTENSION_CALLBACK(
+                    self=self,
```
let make sure we use same format for the entire file e.g. `a = b` not `a=b`

jbhul replied:
🫡
Will update

kyxap reviewed (changes requested):
Left some comments, I will try to find time to build and test on my end this week.
Again thanks for MR!

kyxap commented:
Also lets rebase with `main` I've pushed another PR

jbhul committed:
Merge branch 'harmtemolder:main' into main

jbhul commented:
> Also lets rebase with `main` I've pushed another PR

Done

kyxap commented:
Ideas about UI:

1) Maybe we do not need `status` and `result` columns use just one and marge data there?

<img width="579" alt="image" src="https://github.com/user-attachments/assets/a7e70b28-ba01-4452-9523-a5b48cd0797c" />

jbhul committed:
Update gitignore and EXTENSION_CALLBACK case

Adds .temp/ as a ignored directory for my use case of test .py scripts/files and a TODO
EXTENSION_CALLBACK >> extension_callback (convention)
Corrected a=b to a = b for consistency

jbhul committed:
Merge Status Col into Results

Combines status into result col of SyncCompletionDialog

jbhul commented:
> 1. `result`

Addressed in [bd5bf9d](https://github.com/harmtemolder/koreader-calibre-plugin/pull/71/commits/bd5bf9dcc377d690621f7edff19ff26905d12865)

kyxap commented:
Also seems like bug to me, column has data and also `NONE` can you take a look? 

<img width="1010" alt="image" src="https://github.com/user-attachments/assets/23261aef-5ff6-4091-bc68-d29c3893691b" />

jbhul commented:
> Also seems like bug to me, column has data and also `NONE` can you take a look?
> 
> <img alt="image" width="1010" src="https://private-user-images.githubusercontent.com/3080529/438540222-23261aef-5ff6-4091-bc68-d29c3893691b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4OTgzNjMsIm5iZiI6MTc0NTg5ODA2MywicGF0aCI6Ii8zMDgwNTI5LzQzODU0MDIyMi0yMzI2MWFlZi01ZmY2LTQwOTEtYmM2OC1kMjljMzg5MzY5MWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjlUMDM0MTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWE1ODFjYjdkNGZkNDI3OGI3Zjc1YzEyMjhiZTc0NDg5ZWNiZGNjZmI4M2M4ZDgzZGFiMWE4MTZhNjk1YTZjZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JryHRNUTyRuqWSBUNGAncZrJsztd1orvGZWL-NY6RJw">

The table now shows changes in the format "A >> B"
So the highlighted example was empty before sync

kyxap commented on action.py:
```diff
@@ -726,196 +805,564 @@ def sync_missing_sidecars_to_koreader(self):
-                    **result,
-                    'book_uuid': book_uuid,
-                    'sidecar_path': sidecar_path,
-                    # too much data, hard to read for user
```
it still seems like there can be a lot of data which sometime can't be displayed at all, do still want to have it there?

<img width="798" alt="image" src="https://github.com/user-attachments/assets/cf2154c9-8340-48a8-b9f0-ce33b3dd6028" />


Also how hard it can be to add additional config what columns on include in the report? (Can be done separate PR, or as feature request)

jbhul replied:
What col is this? this just seems blank.
In the config there def could be a checkbox but I don't think that would be as helpful as maybe reducing font size in the hover.
Realistically I think it's mainly the raw sidecar that does this, maybe bookmarks? TBH I don't think we need to show the raw in the table, esp not diff.

kyxap commented on action.py:
Time to time I'm getting error while plugin doing books sync somewhere after 5-6 books it could be because I have my library db stored on network disk which is not recommended but I never had an issue before, this started to happen when I've enabled most of the column to sync

```bash
calibre, version 8.3.0
ERROR: Unhandled exception: <b>CantOpenError</b>:CantOpenError: unable to open database file

calibre 8.3  embedded-python: True
macOS-15.4.1-arm64-arm-64bit Darwin ('64bit', '')
('Darwin', '24.4.0', 'Darwin Kernel Version 24.4.0: Fri Apr 11 18:33:47 PDT 2025; root:xnu-11417.101.15~117/RELEASE_ARM64_T6000')
Python 3.11.12
OSX: ('15.4.1', ('', '', ''), 'arm64')
Interface language: None
EXE path: /Applications/calibre.app/Contents/MacOS/calibre
Successfully initialized third party plugins: KOReader Sync (0, 6, 8)
Traceback (most recent call last):
  File "calibre/db/cache.py", line 1916, in set_metadata
  File "calibre/db/cache.py", line 1852, in protected_set_field
  File "calibre/db/cache.py", line 1830, in set_field
  File "calibre/db/cache.py", line 1643, in set_field
  File "calibre/db/write.py", line 572, in set_books
  File "calibre/db/write.py", line 238, in one_one_in_other
  File "calibre/db/backend.py", line 1183, in executemany
  File "src/cursor.c", line 240, in resetcursor
apsw.CantOpenError: CantOpenError: unable to open database file

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.koreader.action", line 1167, in run
    operation_status, result = self.action.update_metadata(
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "calibre_plugins.koreader.action", line 595, in update_metadata
    db.set_metadata(
  File "calibre/db/cache.py", line 85, in call_func_with_lock
  File "calibre/db/cache.py", line 1926, in set_metadata
  File "calibre/db/cache.py", line 326, in reload_from_db
  File "calibre/db/backend.py", line 113, in load_from_db
  File "calibre/db/backend.py", line 410, in get
  File "src/cursor.c", line 240, in resetcursor
apsw.CantOpenError: CantOpenError: unable to open database file
```

jbhul replied:
hmmm not too sure here. I don't get this error with 100+ books but hard to say exactly what the problem is with the logs, they aren't super descriptive. My guess is something locking related? Maybe the changes that reduced db = self.gui.current_db.new_api helps here?

kyxap replied:
still have this issue, but I think happening when it on my remote host I think its fine at this time

kyxap commented on config.py:
```diff
@@ -47,136 +55,255 @@
+                'Restart calibre to apply this setting',
+    },
+    'checkbox_enable_scheduled_progressync': {
+        'config_label': 'Enable Daily ProgressSync',
```
Lets make it `[ ] Daily ProgressSync` there no need to have extra `Enable` if checkbox represents boolean value

kyxap commented on action.py:
```diff
@@ -726,196 +805,564 @@ def sync_missing_sidecars_to_koreader(self):
+                )
+                res['results'].sort(key=lambda row: (not row.get('error', False), -len(row))) # Sort by if error, then # of changes
+                if res['num_success'] > 0 and res['num_fail'] == 0:
+                    SyncCompletionDialog(
```
Currently this window is always on top which make it hard to look into calibre for testing/debugging can this be a regular window which allow you simple switch to main Calibre window?

jbhul replied:
Yep, Exec to show iirc

jbhul replied:
1323 in action self.show()

kyxap commented on action.py:
```diff
@@ -726,196 +805,564 @@ def sync_missing_sidecars_to_koreader(self):
+        custom_columns = sorted(h for h in all_headers 
+                               if h not in ('book_uuid', 'result', 'error'))
+
+        if 'book_uuid' in all_headers:
```
have UUID in table is good but I find that it's hard to quickly look which exactly book is it can we book name\and maybe author?

jbhul replied:
I agree here, Not too sure if that data is easily accessible but will check

jbhul replied:
Just added title here

kyxap commented on action.py:
```diff
@@ -726,196 +805,564 @@ def sync_missing_sidecars_to_koreader(self):
-
+        
+        'Get list of books with MD5 column'
+        db = self.gui.current_db.new_api
```
Now we use `db = self.gui.current_db.new_api` 4 times in the code I do not think this can affect the performance but maybe we can avoid duplication and do the null check before reassign the `db`? (not sure if it any improvement in python at all but no dublication is good)

jbhul replied:
Updating update_metadata to use book_id instead of a passed uuid would remove the redundancy. 

Realistically it's called up to twice for any particular call. prog sync declares it once and then each update meta declares it once which is the real issue here.

jbhul replied:
Update: Nope nvm. 
I just updated the update_metadata to include db as passed. 
db = self.gui.current_db.new_api is now only called once whenever any of the menu options are clicked.

kyxap replied:
`self.gui.current_db.new_api` is still uses in 3 places:

![Uploading image.png…]()

kyxap replied:
<img width="652" alt="image" src="https://github.com/user-attachments/assets/ae7e7456-b396-416a-8170-3fbfc52306cc" />

jbhul replied:
Because it's needed in 3 different places, the current plugin uses it twice. Also the # of times it's used in the code doesn't frankly matter, it's the # of times it's used within a specific call and if each time is needed.

For example, if I click progress sync it calls it once for the whole run. Same with sync from koreader to calibre.
What's the alternative here?

The 3rd use is unchanged for calibre to sidecar.

jbhul committed:
Readability and Redundancy Upgrades

Reduces db = self.gui.current_db.new_api instantiation.
Adds title to table for readability.
Switched sync completion dialog from exec_ to show to allow for background manipulation.
Renamed config label for Progress Sync

jbhul commented:
[cafc46d](https://github.com/harmtemolder/koreader-calibre-plugin/pull/71/commits/cafc46dbbb1dc16c76ecee4ac658a2872cb5ff42) addresses most of the requests.
May address the db access error but needs to be tested since I can't reproduce.
Need to discuss the result table a bit before tweaks there.

kyxap commented on action.py:
```diff
@@ -488,6 +551,20 @@ def update_metadata(self, uuid, keys_values_to_update):
+                updateLog = self.extension_callback(
+                    self = self,
+                    metadata = metadata,
+                    keys_values_to_update = keys_values_to_update,
```
btw what IDE you are using and what style format? I've just realized that Intellij auto-format this into `keys_values_to_update=keys_values_to_update,` would nice to have same format and just apply auto-format for the entire project.

jbhul replied:
I used VS Code

kyxap replied:
can you apply auto-format and push the changes if they look good to you?

jbhul replied:
Done

jbhul committed:
Autoformatted

Autoformatted per request

kyxap commented:
Linked all feature requests to this PR and left 1 comment regarding db, lets look into it and we can merge and I will pre the release.

jbhul commented:
> Linked all feature requests to this PR and left 1 comment regarding db, lets look into it and we can merge and I will pre the release.

Got it, I responded to that comment and I think it's a non-issue. It's not problematic in my other plugin where there's a similar implementation.
